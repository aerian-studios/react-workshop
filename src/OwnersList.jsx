import "../__mocks__/fetch"; // we don't really have an API!
import React, { useState, useEffect } from "react";
import { extractDataByOwner, getData } from "./dataModel";

export const generateList = (owners) =>
    Object.keys(owners).map((owner) => {
        const elements = owners[owner];
        return (
            <tr key={owner}>
                <td data-testid="owner">{owner}:</td>
                <td>
                    {elements.map((element) => (
                        <span key={element}>{element}</span>
                    ))}
                </td>
            </tr>
        );
    });

export const OwnersList = () => {
    const [owners, setOwners] = useState(null);

    useEffect(() => {
        const getOwnerData = async () => {
            const owners = await getData();
            setOwners(extractDataByOwner(owners));
        };
        getOwnerData();
    }, []);

    return owners ? (
        <table>
            <thead>
                <tr>
                    <th>Owner</th>
                    <th>Elements</th>
                </tr>
            </thead>
            <tbody>{generateList(owners)}</tbody>
        </table>
    ) : (
        <p>Loading...</p>
    );
};
