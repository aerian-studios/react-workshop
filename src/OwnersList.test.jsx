import React from "react";

import {
    render,
    cleanup,
    fireEvent,
    getByText,
    getAllByTestId,
} from "@testing-library/react";

import { OwnersList } from "./OwnersList";

describe("OwnersList", () => {
    afterEach(cleanup);
    it("renders a table correctly", async () => {
        const { container, findByText } = render(<OwnersList />);
        expect(container).toMatchSnapshot();
        await findByText("Owner");
        expect(container).toMatchSnapshot();
    });

    it("reverses the order of the table when clicking on table headers", async () => {
        const { container, findByText } = render(<OwnersList />);
        await findByText("Owner");
        expect(container).toMatchSnapshot();
        let owners = getAllByTestId(container, "owner");
        expect(owners.length).toBe(3);
        expect(owners[0].textContent).toBe("bob:");
        expect(owners[1].textContent).toBe("fred:");
        expect(owners[2].textContent).toBe("kevin:");

        fireEvent.click(getByText(container, "Owner"));
        // expect(container).toMatchSnapshot();
        owners = getAllByTestId(container, "owner");
        // expect(owners[0].textContent).toBe("kevin:");
        // expect(owners[1].textContent).toBe("fred:");
        // expect(owners[2].textContent).toBe("bob:");
    });
});
