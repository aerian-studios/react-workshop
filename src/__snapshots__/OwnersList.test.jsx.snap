// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OwnersList renders a table correctly 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;

exports[`OwnersList renders a table correctly 2`] = `
<div>
  <table>
    <thead>
      <tr>
        <th>
          Owner
        </th>
        <th>
          Elements
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          data-testid="owner"
        >
          bob
          :
        </td>
        <td>
          <span>
            first_element
          </span>
          <span>
            third_element
          </span>
        </td>
      </tr>
      <tr>
        <td
          data-testid="owner"
        >
          fred
          :
        </td>
        <td>
          <span>
            feedback
          </span>
          <span>
            fifth_element
          </span>
        </td>
      </tr>
      <tr>
        <td
          data-testid="owner"
        >
          kevin
          :
        </td>
        <td>
          <span>
            fourth_element
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`OwnersList reverses the order of the table when clicking on table headers 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th>
          Owner
        </th>
        <th>
          Elements
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          data-testid="owner"
        >
          bob
          :
        </td>
        <td>
          <span>
            first_element
          </span>
          <span>
            third_element
          </span>
        </td>
      </tr>
      <tr>
        <td
          data-testid="owner"
        >
          fred
          :
        </td>
        <td>
          <span>
            feedback
          </span>
          <span>
            fifth_element
          </span>
        </td>
      </tr>
      <tr>
        <td
          data-testid="owner"
        >
          kevin
          :
        </td>
        <td>
          <span>
            fourth_element
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
